sw/device/silicon_creator/lib/sigverify/mod_exp_otbn.c

```diff
@@ -35,6 +35,21 @@ static const otbn_addr_t kOtbnVarRsaInBuf =
 static const otbn_addr_t kOtbnVarRsaM0Inv =
     OTBN_ADDR_T_INIT(run_rsa_verify_3072_rr_modexp, m0inv);
 
+/**
+ * Manually-calculated instruction count range.
+ *
+ * The OTBN modexp implementation is not constant-time (which is okay because
+ * it has no secret inputs). However, all valid control-flow paths should fall
+ * within this range, which is calculated by changing the code to either take
+ * all branches or skip them all.
+ *
+ * IMPORTANT: This may need to be modified if the modexp routine is changed!
```

Comment:
I imagine that it might be cumbersome to implement but ideally we would have some kind of CI test to automatically verify these bounds.

Reply:
We can take a pragmatic approach here and just rely on the tests that we will have for all production keys. As long as they pass we are good.

_If_ we can find the inputs that trigger the right conditions, adding a functional test that compares the instruction counts to the constants here should be trivial. I guess, finding those inputs is the hard part. @jadephilipoom could you please create an issue that captures this? Thanks!

Reply:
Finding inputs that trigger all the right conditions will be extremely difficult if not impossible -- the case where the branch has to be taken in `mont_loop` is very rare, and we would need to find an input that hits that case for all 276 intermediate values on which the function is called. We could try to create a routine that modifies the code to take those branches and then reads the counts, but fully automating that would be a lot of work.

One thing we could do long-term: some (but not all) of the code in the OTBN program is linked to veri-titan's machine model. We could add the instruction count register to their model so it gets properly incremented each time, and add a clause to the function specification that says "instruction count within this range". Then we'd have a formal guarantee that the range is not too strict. However, we'd need to cover the whole OTBN program with the verification to get much use out of this guarantee (right now we could get a range for the `modexp` method but not for loading the data or calculating R^2) -- and it would still need to be manually updated here if it changed in the spec.

Reply:
I see, then should add some sort of a digest check for all the OTBN source code that can affect instruction count to make sure things don't go out of sync?

Reply:
Created https://github.com/lowRISC/opentitan/issues/13185